### ISSUE TITLE
runtime: race condition raised with parallel tests, panic(nil) and -race

### ISSUE THREAD

**Issue by dolmen** [golang/go#64649]
### Go version

go version go1.21.5 darwin/arm64

### What operating system and processor architecture are you using (`go env`)?

```shell
- Local (Apple M2)
- GitHub Actions (x86)
```

### What did you do?

I have a testsuite for my [github.com/dolmen-go/rendezvous](https://pkg.go.dev/github.com/dolmen-go/rendezvous) package.

```
$ git clone https://github.com/dolmen-go/rendezvous.git
$ cd rendezvous
$ git checkout 44e68d01a216b3755c9ba5dc75835074c3e5e866
$ go version
go version go1.21.5 darwin/arm64
$ go test -v -race -run 'Test(One|Two)' "-gcflags=all=-trimpath=$(go env GOPATH)" github.com/dolmen-go/rendezvous
=== RUN   TestOne
=== PAUSE TestOne
=== RUN   TestTwo
=== PAUSE TestTwo
=== CONT  TestOne
=== CONT  TestTwo
==================
WARNING: DATA RACE
Read at 0x00c000076560 by goroutine 14:
  internal/godebug.(*Setting).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:102 +0x58
  internal/godebug.(*Setting).IncNonDefault-fm()
      <autogenerated>:1 +0x34
  runtime.(*godebugInc).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/runtime.go:111 +0xd0
  github.com/dolmen-go/rendezvous.WaitAll.func1()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:56 +0xec
  github.com/dolmen-go/rendezvous.WaitAll.func2()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:57 +0x44

Previous write at 0x00c000076560 by goroutine 12:
  internal/godebug.New()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:74 +0x40
  internal/godebug.newIncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:203 +0x28
  runtime.(*godebugInc).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/runtime.go:108 +0x84
  github.com/dolmen-go/rendezvous.WaitAll.func1()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:56 +0xec
  github.com/dolmen-go/rendezvous.WaitAll.func2()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:57 +0x44

Goroutine 14 (running) created at:
  github.com/dolmen-go/rendezvous.WaitAll()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:47 +0x210
  github.com/dolmen-go/rendezvous_test.TestTwo()
      src/github.com/dolmen-go/rendezvous/rendezvous_test.go:118 +0xe4
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1595 +0x1b0
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1648 +0x40

Goroutine 12 (running) created at:
  github.com/dolmen-go/rendezvous.WaitAll()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:47 +0x210
  github.com/dolmen-go/rendezvous_test.TestOne()
      src/github.com/dolmen-go/rendezvous/rendezvous_test.go:104 +0x180
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1595 +0x1b0
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1648 +0x40
==================
=== NAME  TestOne
    testing.go:1465: race detected during execution of test
--- FAIL: TestOne (0.00s)
=== NAME  TestTwo
    testing.go:1465: race detected during execution of test
--- FAIL: TestTwo (0.00s)
FAIL
exit status 1
FAIL	github.com/dolmen-go/rendezvous	0.104s
```

Some characteristics that might be relevant:
* The `rendezvous` package launches goroutines in its tests.
* The `rendezvous` testsuite uses parallel testing at top testing level:

```go
func TestOne(t *testing.T) {
        t.Parallel()
        // ...
}

func TestTwo(t *testing.T) {
        t.Parallel()
        // ...
}
```

This issue is also visible on GitHub Actions: https://github.com/dolmen-go/rendezvous/actions/runs/7147035848/job/19465833264#step:4:32

### What did you expect to see?

Test success like with previous Go versions.

### What did you see instead?

Race condition in `runtime`:

```
WARNING: DATA RACE
Read at 0x00c000076560 by goroutine 14:
  internal/godebug.(*Setting).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:102 +0x58
  internal/godebug.(*Setting).IncNonDefault-fm()
      <autogenerated>:1 +0x34
  runtime.(*godebugInc).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/runtime.go:111 +0xd0
  github.com/dolmen-go/rendezvous.WaitAll.func1()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:56 +0xec
  github.com/dolmen-go/rendezvous.WaitAll.func2()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:57 +0x44

Previous write at 0x00c000076560 by goroutine 12:
  internal/godebug.New()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:74 +0x40
  internal/godebug.newIncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:203 +0x28
  runtime.(*godebugInc).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/runtime.go:108 +0x84
  github.com/dolmen-go/rendezvous.WaitAll.func1()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:56 +0xec
  github.com/dolmen-go/rendezvous.WaitAll.func2()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:57 +0x44

```

Link to runtime source: https://github.com/golang/go/blob/go1.21.5/src/runtime/runtime.go#L97

A [comment](https://github.com/golang/go/blob/6018ad99a4a951581b2d846a8ccd6f1d4e74fd11/src/runtime/runtime.go#L104C3-L106C23) says:
```go
		// If other goroutines are racing here, no big deal. One will win,
		// and all the inc functions will be using the same underlying
		// *godebug.Setting.
```

If this is related, I think this is "big deal" as running `go test -race` in CI is quite common. (Calling `t.Parallel()` from top level tests is not so common, this is unfortunate, and this issue blocks from pushing for it)

**Comment by dolmen** [comment]
```console
$ go test -race -run 'Test(One|Two)' "-gcflags=all=-trimpath=$(go env GOPATH)" -count=1 -parallel=1 
PASS
ok  	github.com/dolmen-go/rendezvous	1.271s
$ go test -race -run 'Test(One|Two)' "-gcflags=all=-trimpath=$(go env GOPATH)" -count=1 -parallel=2
==================
WARNING: DATA RACE
Read at 0x00c000076500 by goroutine 14:
  internal/godebug.(*Setting).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:102 +0x58
  internal/godebug.(*Setting).IncNonDefault-fm()
      <autogenerated>:1 +0x34
  runtime.(*godebugInc).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/runtime.go:111 +0xd0
  github.com/dolmen-go/rendezvous.WaitAll.func1()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:56 +0xec
  github.com/dolmen-go/rendezvous.WaitAll.func2()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:57 +0x44

Previous write at 0x00c000076500 by goroutine 12:
  internal/godebug.New()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:74 +0x40
  internal/godebug.newIncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/internal/godebug/godebug.go:203 +0x28
  runtime.(*godebugInc).IncNonDefault()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/runtime.go:108 +0x84
  github.com/dolmen-go/rendezvous.WaitAll.func1()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:56 +0xec
  github.com/dolmen-go/rendezvous.WaitAll.func2()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:57 +0x44

Goroutine 14 (running) created at:
  github.com/dolmen-go/rendezvous.WaitAll()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:47 +0x210
  github.com/dolmen-go/rendezvous_test.TestTwo()
      src/github.com/dolmen-go/rendezvous/rendezvous_test.go:118 +0xe4
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1595 +0x1b0
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1648 +0x40

Goroutine 12 (running) created at:
  github.com/dolmen-go/rendezvous.WaitAll()
      src/github.com/dolmen-go/rendezvous/rendezvous.go:47 +0x210
  github.com/dolmen-go/rendezvous_test.TestOne()
      src/github.com/dolmen-go/rendezvous/rendezvous_test.go:104 +0x180
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1595 +0x1b0
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1648 +0x40
==================
--- FAIL: TestOne (0.00s)
    testing.go:1465: race detected during execution of test
--- FAIL: TestTwo (0.00s)
    testing.go:1465: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/dolmen-go/rendezvous	0.314s
```

**Comment by mauri870** [comment]
cc @golang/runtime

**Comment by prattmic** [comment]
Read is here: https://cs.opensource.google/go/go/+/master:src/internal/godebug/godebug.go;l=102;drc=39effbc105f5c54117a6011af3c48e3c8f14eca9

Write is here: https://cs.opensource.google/go/go/+/master:src/internal/godebug/godebug.go;l=74;drc=39effbc105f5c54117a6011af3c48e3c8f14eca9

The write is the initialization of the struct. It must happen before the read. Indeed, I believe the atomic store at https://cs.opensource.google/go/go/+/release-branch.go1.21:src/runtime/runtime.go;l=109;drc=1a9893a969e0a73dc4f1e48ed40ccaf29ec238a6 is publishing the value. But this is a runtime/internal/atomic value, which doesn't include race annotations. So I think we need to manually annotate that path.

**Comment by mauri870** [comment]
@prattmic I was looking at this the other day and I also think we should instrument it manually. So something like this would suffice? I'm wondering how to come up with a test for this 😅 

```
diff --git a/src/runtime/runtime.go b/src/runtime/runtime.go
index 0829a84e43..30e7226e0a 100644
--- a/src/runtime/runtime.go
+++ b/src/runtime/runtime.go
@@ -161,6 +161,7 @@ type godebugInc struct {
 }
 
 func (g *godebugInc) IncNonDefault() {
+	var racer int
 	inc := g.inc.Load()
 	if inc == nil {
 		newInc := godebugNewIncNonDefault.Load()
@@ -172,7 +173,9 @@ func (g *godebugInc) IncNonDefault() {
 		// *godebug.Setting.
 		inc = new(func())
 		*inc = (*newInc)(g.name)
+		raceacquire(unsafe.Pointer(&racer))
 		g.inc.Store(inc)
+		racerelease(unsafe.Pointer(&racer))
 	}
 	(*inc)()
 }
```

**Comment by dolmen** [comment]
Has anyone analyzed why this happens with parallel testing? What Go debug events metric is incremented in that case?

Is there a workaround besides:
* not using parallel tests
* not using the race detector when parallel testing is enabled: `go test -parallel=2`
* not using parallel tests when the race detector is not enabled: `go test -race -parallel=1`
All those solutions are unfortunately about avoiding parallel testing.

**Comment by dolmen** [comment]
As `godebugInc` is about incrementing a metric, I did more investigation by dumping the metrics that are effectively incremented:

```go
package rendezvous_test

import (
	"fmt"
	"os"
	"runtime/metrics"
	"strings"
	"testing"
)

func TestMain(m *testing.M) {
	// call flag.Parse() here if TestMain uses flags
	exitCode := m.Run()

	descs := metrics.All()
	var godebugMetrics []metrics.Sample
	for i := range descs {
		if strings.HasPrefix(descs[i].Name, "/godebug/") && strings.HasSuffix(descs[i].Name, ":events") {
			godebugMetrics = append(godebugMetrics, metrics.Sample{Name: descs[i].Name})
		}
	}
	metrics.Read(godebugMetrics)
	for i := range godebugMetrics {
		if godebugMetrics[i].Value.Uint64() > 0 {
			fmt.Printf("%s: %v\n", godebugMetrics[i].Name, godebugMetrics[i].Value.Uint64())
		}
	}

	os.Exit(exitCode)
}
```

The result:
```console
go test -v -parallel 10 -run TestOne
=== RUN   TestOne
=== PAUSE TestOne
=== CONT  TestOne
--- PASS: TestOne (0.00s)
PASS
/godebug/non-default-behavior/panicnil:events: 1
ok  	github.com/dolmen-go/rendezvous	0.096s
```

I have in fact [a test](https://github.com/dolmen-go/rendezvous/blob/8f5e644b7652ada9063da49a43342a94109091b2/rendezvous_test.go#L72) that runs `panic(nil)` on purpose. I have to check if I still want that test.

**Comment by dolmen** [comment]
My final workaround is to [drop](https://github.com/dolmen-go/rendezvous/commit/d9e979043700a9c6369106d1b113c9d8fb1f5364) the `panic(nil)` in my tests as it fails on Go 1.21 and is no longer relevant with Go 1.21+.

**Comment by dr2chase** [comment]
I'm having no luck with a reproducer, this is on Darwin amd64, 1.21.5, 1.21.3, also tip.  Here's my current attempt, I'm about to try this on darwin/arm64.  To get an event count I have to specify 1.20 in go.mod, or `GODEBUG=panicnil=1`, but in no case have I been able to tickle a race failure.  (I absolutely believe this is a bug, but a small test would be darn nice.)
```
package parpanic_test

import (
	"fmt"
	"os"
	"runtime/metrics"
	"strings"
	"testing"
)

//go:noinline
func panicNil() (e any) {
	defer func() {
		e = recover()
	}()
	panic(e)
}

var tests = []string{"ant", "bat", "cat", "dog", "emu", "fox", "gnu", "hen"}

func TestMany(t *testing.T) {
	t.Parallel()
	for _, tc := range tests {
		t.Run(tc, func(t *testing.T) {
			t.Log(panicNil())
		})
	}

}

func TestMain(m *testing.M) {
	// call flag.Parse() here if TestMain uses flags
	exitCode := m.Run()
	descs := metrics.All()
	var godebugMetrics []metrics.Sample
	for i := range descs {
		if strings.HasPrefix(descs[i].Name, "/godebug/") && strings.HasSuffix(descs[i].Name, ":events") {
			godebugMetrics = append(godebugMetrics, metrics.Sample{Name: descs[i].Name})
		}
	}
	metrics.Read(godebugMetrics)
	for i := range godebugMetrics {
		if godebugMetrics[i].Value.Uint64() > 0 {
			fmt.Printf("%s: %v\n", godebugMetrics[i].Name, godebugMetrics[i].Value.Uint64())
		}
	}
	os.Exit(exitCode)
}
```

**Comment by dolmen** [comment]
Here is my small reproducer:

```
package main_test

import "testing"

func Test(t *testing.T) {
	t.Run("One", func(t *testing.T) {
		t.Parallel()
		defer func() {
			recover()
		}()
		panic(nil)
	})
	t.Run("Two", func(t *testing.T) {
		t.Parallel()
		defer func() {
			recover()
		}()
		panic(nil)
	})
}
```

```console
$ GODEBUG=panicnil=1 go test -run 'Test$' -race
```

**Comment by dolmen** [comment]
@dr2chase Your `TestMany` executes in parallel, but there is no parallel test to run. The subtests are not run in parallel.

**Comment by dr2chase** [comment]
@gopherbot please consider this for backport to 1.21.  It broke someone's integration tests, and since `GODEBUG=panicnil` (or go.mod version = 1.20) is a compatibility feature added to 1.21, it is a regression, if you look at it right.

**Comment by dolmen** [comment]
The fix has been released with Go 1.21.6 (see #64757).
